docs/apache-airflow/howto/create-custom-decorator.rst

```diff
@@ -0,0 +1,118 @@
+By creating a ``FooDecoratedOperator`` that inherits from ``FooOperator`` and
+``airflow.decorators.base.DecoratedOperator``, Airflow will supply much of the needed functionality required to treat
+your new class as a taskflow native class.
+
+2. Create a ``foo_task`` function
+
+Once you have your decorated class, create a function that takes arguments ``python_callable``\, ``multiple_outputs``\,
+and ``kwargs``\. This function will use the ``airflow.decorators.base.task_decorator_factory`` function to convert
+the new ``FooDecoratedOperator`` into a TaskFlow function decorator!
+
+.. code-block:: python
+
+   def foo_task(
+       python_callable: Optional[Callable] = None,
+       multiple_outputs: Optional[bool] = None,
+       **kwargs
+   ):
+       return task_decorator_factory(
+           python_callable=python_callable,
+           multiple_outputs=multiple_outputs,
+           decorated_operator_class=FooDecoratedOperator,
+           **kwargs,
+       )
+
+3. Register your new decorator in the provider.yaml of your provider
+
+Finally, add a key-value of ``decorator-name``:``path-to-function`` to your provider.yaml. When Airflow starts, the
+``ProviderManager`` class will automatically import this value and ``task.decorator-name`` will work as a new
+decorator!
+
+.. code-block:: yaml
+
+   package-name: apache-airflow-providers-docker
+   name: Docker
+   description: |
+       `Docker <https://docs.docker.com/install/>`__
+
+   task-decorators:
+       docker: airflow.providers.docker.operators.docker.docker_decorator
+
+
+4. (Optional) Create a Mixin class so that your decorator will show up in your IDE's autocomplete
+
+For better or worse, Python IDEs can not autocomplete dynamically
+generated methods (see `here <https://intellij-support.jetbrains.com/hc/en-us/community/posts/115000665110-auto-completion-for-dynamic-module-attributes-in-python>`_).
+
+To get around this, we had to find a solution that was "best possible." IDEs will only allow typing
+through stub files, but we wanted to avoid any situation where a user would update their provider and the autocomplete
+would be out of sync with the provider's actual parameters.
+
+To hack around this problem, we found that you could extend the ``_TaskDecorator`` class in the ``__init__.pyi`` file
+and the correct autocomplete will show up in the IDE.
+
+To correctly implement this, please take the following steps:
+
+Create a ``Mixin`` class for your decorator
+
+Mixin classes are classes in python that tell the python interpreter that python can import them at any time.
+Because they are not dependent on other classes, Mixin classes are great for multiple inheritance.
+
+In the DockerDecorator we created a Mixin class that looks like this
+
+.. exampleinclude:: ../howto/docker_decorator.py
+    :language: python
+    :start-after: [START decoratormixin]
+    :end-before: [END decoratormixin]
```

Comment:
Rather than including a whole copy of docker_decorator.py here (that will get out of date) can we instead point at the actual airflow/providers/docker/decorators/docker.py?